--- C/Compress/Lzma/LzmaStateTest.c.orig	2005-08-02 10:28:00.000000000 +0200
+++ C/Compress/Lzma/LzmaStateTest.c	2008-04-18 20:22:55.000000000 +0200
@@ -10,6 +10,14 @@ This file is part of LZMA SDK 4.26 (2005
 #include <stdlib.h>
 #include <string.h>
 
+#if defined(_WIN32) || defined(OS2) || defined(MSDOS)
+#include <fcntl.h>
+#include <io.h>
+#define MY_SET_BINARY_MODE(file) _setmode(_fileno(file), O_BINARY)
+#else
+#define MY_SET_BINARY_MODE(file)
+#endif
+
 #include "LzmaStateDecode.h"
 
 const char *kCantReadMessage = "Can not read input file";
@@ -173,7 +181,12 @@ int main2(int numArgs, const char *args[
 
   if (numArgs > 2)
   {
-    outFile = fopen(args[2], "wb+");
+    if (strncmp(args[2],"-",1)==0) {
+      outFile = stdout;
+      MY_SET_BINARY_MODE(stdout);
+    } else {    
+      outFile = fopen(args[2], "wb+");
+    }
     if (outFile == 0)
       return PrintError(rs, "Can not open output file");
   }
--- C/Compress/Lzma/LzmaStateTest.c.orig	2005-08-02 10:28:00.000000000 +0200
+++ C/Compress/Lzma/LzmaStateTest.c	2008-04-18 20:22:55.000000000 +0200
@@ -10,6 +10,14 @@ This file is part of LZMA SDK 4.26 (2005
 #include <stdlib.h>
 #include <string.h>
 
+#if defined(_WIN32) || defined(OS2) || defined(MSDOS)
+#include <fcntl.h>
+#include <io.h>
+#define MY_SET_BINARY_MODE(file) _setmode(_fileno(file), O_BINARY)
+#else
+#define MY_SET_BINARY_MODE(file)
+#endif
+
 #include "LzmaStateDecode.h"
 
 const char *kCantReadMessage = "Can not read input file";
@@ -173,7 +181,12 @@ int main2(int numArgs, const char *args[
 
   if (numArgs > 2)
   {
-    outFile = fopen(args[2], "wb+");
+    if (strncmp(args[2],"-",1)==0) {
+      outFile = stdout;
+      MY_SET_BINARY_MODE(stdout);
+    } else {    
+      outFile = fopen(args[2], "wb+");
+    }
     if (outFile == 0)
       return PrintError(rs, "Can not open output file");
   }
